scripts/usb_reset.py

```diff
@@ -14,46 +14,266 @@
 # along with this program; if not, write to the Free Software Foundation, Inc.,
 # 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 #
-# ./usb_reset.py 1-2
-# reset USB device 1-2
+# attach
+# ./usb_reset.py attach device -d dom-id -p pid [-r]
+# ./usb_reset.py attach 2-2 -d 12 -p 4130
+# ./usb_reset.py attach 2-2 -d 12 -p 4130 -r
+# 1. reset device
+# if without -r, do step 2~4
+# 2. if it's the first USB device to pass-through
+#      a) bind mount /dev /sys in chroot directory (/var/xen/qemu/<domid))
+#      b) create new cgroup devices:/qemu-<domid>,
+#      c) blacklist all and add default device whitelist,
+#      d) join current qemu process to this cgourp
+# 3. set device file uid/gid to (qemu_base + dom-id)
+# 4. add current device to whitelist
+#
+# detach
+# ./usb_reset.py detach device -d dom-id -i uid gid
+# ./usb_reset.py detach 2-2 -d 12 -i 0 0
+# 1. restore device file owner to uid/gid
+# 2. remove current device from whitelist
+#
+# cleanup
+# ./usb_reset.py cleanup -d dom-id
+# ./usb_reset.py cleanup -d 12
+# 1.remove the cgroup if one has been created.
+# 2.umount /dev, /sys from chroot directory if they are mounted.
 
+import argparse
+import ctypes
+import ctypes.util
+import errno
 import fcntl
+import grp
 import xcp.logger as log
 import logging
+import os
+import pwd
 import re
-import sys
-import traceback
+from stat import S_ISCHR, S_ISBLK
+
+
+def parse_arg():
+    parser = argparse.ArgumentParser(
+        description="script to attach, detach, cleanup for USB pass-through")
+    subparsers = parser.add_subparsers(help="command", dest="command")
+
+    attach = subparsers.add_parser("attach", help="attach a USB device")
+    attach.add_argument("device", help="the target usb device")
+    attach.add_argument("-d", dest="dom_id", type=int, required=True,
+                        help="specify the dom_id of the VM")
+    attach.add_argument("-p", dest="pid", type=int, required=True,
+                        help="the process id of QEMU")
+    attach.add_argument("-r", dest="reset_only", type=int,
+                        help="reset device only, for privileged mode")
+
+    detach = subparsers.add_parser("detach", help="detach a USB device")
+    detach.add_argument("device", help="the target usb device")
+    detach.add_argument("-d", dest="dom_id", type=int, required=True,
+                        help="specify the dom_id of the VM")
+    detach.add_argument("-i", dest="id", type=int, nargs=2, required=True,
+                        help="restore ownership to uid gid")
+
+    cleanup = subparsers.add_parser("cleanup", help="clean up chroot directory")
+    cleanup.add_argument("-d", dest="dom_id", type=int, required=True,
+                         help="specify the dom_id of the VM")
+
+    return parser.parse_args()
 
-#USBDEVFS_RESET _IO('U', 20)
-USBDEVFS_RESET = (ord('U') << 8) | 20
 
+def get_root_dir(dom_id):
+    return "/var/xen/qemu/{}/".format(dom_id)
```

Comment:
I'm not sure about Python conventions but it might be more natural to not have a trailing `/`. This way, appending a path would become `get_root_dir(id) + "/some/path"` rather than `get_root_dir(id) + "some/path"`.

Reply:
Thanks. It's easy for me to assume abc/ as a dir, and /edf probably a file.

Reply:
done.